src/lib/operators/window_function_evaluator.cpp

```diff
@@ -0,0 +1,607 @@
+    std::string error_message = "Unsupported frame type and order-by column type combination: ";
+    error_message += [&]() {
+      using std::literals::string_view_literals::operator""sv;
+
+      switch (frame_type) {
+        case FrameType::Rows:
+          return "Rows"sv;
+        case FrameType::Range:
+          return "Range"sv;
+        case FrameType::Groups:
+          return "Groups"sv;
+        default:
+          return ""sv;
+      }
+    }();
+    error_message += " ";
+    error_message += data_type_to_string.left.at(data_type_from_type<OrderByColumnType>());
+    Fail(error_message);
+  }
+};
+
+template <typename OrderByColumnType>
+struct WindowBoundCalculator<FrameType::Rows, OrderByColumnType> {
+  static QueryRange calculate_window_bounds(std::span<const RelevantRowInformation> partition, uint64_t tuple_index,
+                                            const FrameDescription& frame) {
+    const auto start = frame.start.unbounded ? 0 : clamped_sub<uint64_t>(tuple_index, frame.start.offset, 0);
+    const auto end = frame.end.unbounded ? partition.size()
+                                         : clamped_add<uint64_t>(tuple_index, frame.end.offset + 1, partition.size());
+    return {start, end};
+  }
+};
+
+template <>
+struct WindowBoundCalculator<FrameType::Range, NoOrderByColumn> {
+  static QueryRange calculate_window_bounds(std::span<const RelevantRowInformation> partition,
+                                            [[maybe_unused]] uint64_t tuple_index,
+                                            [[maybe_unused]] const FrameDescription& frame) {
+    return {0, partition.size()};
+  }
+};
+
+template <typename OrderByColumnType>
+  requires std::is_integral_v<OrderByColumnType>
+struct WindowBoundCalculator<FrameType::Range, OrderByColumnType> {
+  static QueryRange calculate_window_bounds(std::span<const RelevantRowInformation> partition, uint64_t tuple_index,
+                                            const FrameDescription& frame) {
+    const auto current_value = as_optional<OrderByColumnType>(partition[tuple_index].order_values[0]);
+
+    // We sort null-first
+    const auto end_of_null_peer_group_it =
+        std::ranges::partition_point(partition, [&](const auto& row) { return variant_is_null(row.order_values[0]); });
+    const auto end_of_null_peer_group = std::distance(partition.begin(), end_of_null_peer_group_it);
+    const auto non_null_values = partition.subspan(end_of_null_peer_group);
+
+    const auto start = [&]() -> uint64_t {
+      if (frame.start.unbounded) {
+        return 0;
+      }
+      if (!current_value) {
+        return 0;
+      }
+      const auto it = std::ranges::partition_point(non_null_values, [&](const auto& row) {
+        return std::cmp_less(get<OrderByColumnType>(row.order_values[0]) + frame.start.offset, *current_value);
+      });
+      return end_of_null_peer_group + std::distance(non_null_values.begin(), it);
```

Comment:
Actually, I think you are not allowed to pass iterators of different origin which is why we did it like this. ~~We could convert both iterators to pointers (which we can then use as if they are from the same span, because they are) and calculate their distance~~ (actually we can't, because we can't dereference the `end` iterator), at the time of writing we felt that this was more clean though. What do you think?